src/pyfreeleh/row/gsheet.py

```diff
@@ -1,412 +1,458 @@
-        return result
+        return raw_values
 
 
 class UpdateStmt:
-    INDICES_FORMULA_TEMPLATE = '=JOIN(",", ARRAYFORMULA(QUERY({{{data_range}, ROW({data_range})}}, "{query}")))'
-    ROW_IDX_FIELD = "_idx"
-
     def __init__(
         self,
-        wrapper: GoogleSheetWrapper,
-        spreadsheet_id: str,
-        sheet_name: str,
-        scratchpad_cell: A1Range,
-        columns: List[str],
+        sheet_session: GoogleSheetSession,
+        mapper: FieldColumnMapper,
         update_values: Dict[str, str],
     ):
-        self._val = update_values
-        self._wrapper = wrapper
-        self._sheet_name = sheet_name
-        self._spreadsheet_id = spreadsheet_id
-        self._scratchpad_cell = scratchpad_cell
-        self._columns = columns
+        self._sheet_session = sheet_session
+        self._mapper = mapper
+        self._update_values = update_values
 
-        c1_col_mapping = get_col_idx_column_mapping(self._columns + [self.ROW_IDX_FIELD])
-        self._query = QueryBuilder(c1_col_mapping)
+        self._query = GoogleSheetQueryBuilder()
 
     def where(self, condition: str, *args: Any) -> "UpdateStmt":
-        self._query.where(condition, *args)
+        """Filter the rows that we're going to update.
+
+        The given `condition` will be used as the WHERE clause on the final query. You can use `"?"` placeholder
+        inside the condition and will be replaced with the actual value given in the `*args` variadic parameter
+        based on their appearance ordering.
+
+        Args:
+            condition: conditions of the data that we're going to update.
+            *args: list of arguments that will be used to fill in the placeholders in the given `condition`.
+
+        Returns:
+            UpdateStmt: the delete statement with the given WHERE condition applied.
+
+        Examples:
+            To apply "WHERE age > 10" filter on the update statement:
+
+            >> store.update({"name": "cat"}).where("age > ?", 10).execute()
+            10
+        """
+        mapped_condition = map_conditions(self._mapper, condition)
+        self._query.where(mapped_condition, *args)
         return self
 
     def execute(self) -> int:
-        data_range = A1Range(self._sheet_name, A1CellSelector.from_rc(1, 2), A1CellSelector.from_rc(len(self._columns)))
-        query = self._query.build_select([self.ROW_IDX_FIELD]).replace('"', '""')
-        formula = self.INDICES_FORMULA_TEMPLATE.format(data_range=str(data_range), query=query)
+        """Execute the update statement.
 
-        result = self._wrapper.update_rows(self._spreadsheet_id, self._scratchpad_cell, [[formula]])
-        update_candidate_indices = [int(idx) for idx in result.updated_values[0][0].split(",") if idx]
+        Returns:
+            int: the number of updated rows.
+        """
+        # TODO(fata.nugraha): assumption pk is in the first cell
+        affected_rows = self._sheet_session.query(self._query.build_select(["A"]))
```

Comment:
You should do something like `SELECT _rid` instead and let the query builder deal with the mapping automatically just like a normal SELECT statement.

Reply:
Actually I moved out the translation layer outside the QueryBuilder so it doesnt have to be aware of this mapping.

Currently the stmt acts as this translation layer.

Reply:
Let's not have another layer, instead let's use the `self._mapper.to_column("_rid")` to do the mapping (instead of creating another unnecessary layer for now).